apiserver/clusterservice/scaleimpl.go

```diff
@@ -292,49 +334,179 @@ func ScaleDown(deleteData bool, clusterName, replicaName, ns string) msgs.ScaleD
 	deleteBackups := false
 	isReplica := true
 	isBackup := false
 	taskName := replicaName + "-rmdata"
-	clusterPGHAScope := cluster.ObjectMeta.Labels[config.LABEL_PGHA_SCOPE]
 	err = apiserver.CreateRMDataTask(clusterName, replicaName, taskName, deleteBackups, deleteData, isReplica, isBackup, ns, clusterPGHAScope)
 	if err != nil {
 		response.Status.Code = msgs.Error
 		response.Status.Msg = err.Error()
 		return response
 	}
 
-	response.Results = append(response.Results, "deleted Pgreplica "+replicaName)
+	response.Results = append(response.Results, "deleted replica "+replicaName)
 	return response
 }
+
+// scaleClusterDeployments scales all deployments for a cluster to the number of replicas
+// specified using the 'replicas' parameter.  This is typically used to scale-up or down the
+// primary deployment and any supporting services (pgBackRest and pgBouncer) when shutting down
+// or starting up the cluster due to a scale or scale-down request.
+func scaleClusterDeployments(cluster crv1.Pgcluster, replicas int) (clusterServices []string,
+	err error) {
+
+	clusterName := cluster.Name
+	namespace := cluster.Namespace
+	// Get *all* remaining deployments for the cluster.  This includes the deployment for the
+	// primary, as well as the pgBackRest repo and any optional deployment (e.g. pgBouncer)
+	var deploymentList *v1.DeploymentList
+	selector := fmt.Sprintf("%s=%s", config.LABEL_PG_CLUSTER, clusterName)
+	if deploymentList, err = kubeapi.GetDeployments(apiserver.Clientset, selector,
+		namespace); err != nil {
+		return nil, err
+	}
+
+	for _, deployment := range deploymentList.Items {
+		log.Debugf("scaling deployment %s to %d for cluster %s", deployment.Name, replicas,
+			clusterName)
+		// scale the deployment
+		if err := kubeapi.ScaleDeployment(apiserver.Clientset, deployment, replicas); err != nil {
+			return nil, err
+		}
+		// determine if the deployment is a supporting service (pgBackRest, pgBouncer, etc.)
+		// and capture the name of the service for display in the response below
+		switch {
+		case deployment.Labels[config.LABEL_PGBOUNCER] == "true":
+			clusterServices = append(clusterServices, "pgBouncer")
+		case deployment.Labels[config.LABEL_PGO_BACKREST_REPO] == "true":
+			clusterServices = append(clusterServices, "pgBackRest")
+		}
+	}
+	return clusterServices, nil
+}
+
+func publishClusterShutdown(cluster *crv1.Pgcluster) error {
+
+	clusterName := cluster.Name
+
+	//capture the cluster creation event
+	topics := make([]string, 1)
+	topics[0] = events.EventTopicCluster
+
+	f := events.EventShutdownClusterFormat{
+		EventHeader: events.EventHeader{
+			Namespace: cluster.Namespace,
+			Username:  cluster.Spec.UserLabels[config.LABEL_PGOUSER],
+			Topic:     topics,
+			Timestamp: time.Now(),
+			EventType: events.EventShutdownCluster,
+		},
+		Clustername: clusterName,
+	}
+
+	if err := events.Publish(f); err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
```

Comment:
I believe if said guidance about the events holds, we can remove this :smile:

Reply:
Now that shutdowns occur asynchronously with a pgcluster update, the shutdown event remains but has been moved from the API server to the Operator.